### ISSUE TITLE
Encrypt data at rest

### ISSUE THREAD

**Issue by voidus** [voidus/coronaSignIn#21]
It would be nice to encrypt the data that is saved in the database.

The threat model here is that someone will either a) get a copy of the database or b) take control of the machines running the service and/or the database.

## Goals

1. Prevent disclosure of records in all cases.
2. Prevent copy of records to different days, i.e. we don't want an adversary to be able to say that everyone who was at the place on saturday also has been there on sunday.
  1. *Iff application hasn't been compromised*: Prevent insertion of chosen data on past days. For the current day, the adversary can just use the form, and if the app server has been compromised the adversary will be able to do everything the app can do.

### Other requirements

1. The data still has to be deleteable by date, so that needs to be unencrypted

### Non-Goals

1. Detect deletion of records

## Algorithm requirements

1. We want the app to create records that cannot be read by it, so we need a *public-key component*.
2. We want the data not to be manipulated, so we need *Authenticated Encryption*
3. We need the date to be authenticated too, but not encrypted, so we need *Authenticated Encryption with Additional Data* [wikipedia](https://en.wikipedia.org/wiki/Authenticated_encryption)

### Preparation
1. We start by generating a a key-pair offline. The secret key is used for decryption and never touches the servers, while the public key is securely made available to the application. *It must not be disclosed to allow Goal 2.1*.

## Encryption
1 We use [Integrated Encryption Scheme (IES)](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme) to randomly derive bytes which we will call SHARED_SECRET_SOURCE. This will also generate some IES_DATA that is needed for decryption and thus has to be stored with the record.
  * I propose using elliptic-curve IES. This allows smaller key sizes, so they could theoretically be stored on paper instead of digitally, but it's not important
    * I propose we use [Curve25519](https://en.wikipedia.org/wiki/Curve25519). It seems to be popular and mostly independent of big government institutions.
2. From SHARED_SECRET_SOURCE, we derive a shared secret (let's call it SHARED_SECRET) using a [Key Derivation Function (KDF)](https://en.wikipedia.org/wiki/Key_derivation_function)
  * This is necessary because the SHARED_SECRET_SOURCE might not be the correct size. Regarding randomness, we can rely on the IES to do it's job. It is my understanding that this is what [HKDF](https://en.wikipedia.org/wiki/HKDF) was designed to do. Since we already have cryptographically strong material (I think), we don't need an additional salt.
3. Using this SHARED_SECRET as a key, we generate a random NONCE and use ChaCha20Poly1305 (A combination of the [ChaCha Cipher](https://en.wikipedia.org/wiki/Salsa20#ChaCha_variant) and the [Poly1305 Message Authentication Code with Associated Data](https://en.wikipedia.org/wiki/Poly1305)).
  * We store the unencrypted date together with the record and also use it as Associated Data. This means that the Poly1305-MAC will fail to validate if an encrypted record is copied over to another date, thus allowing *Goal 2*
  * With XChaCha20Poly1305, the NONCE is long enough (20 bytes) to be randomly generated, but XChaCha20Poly1305 implementations aren't as common as the older, more common ChaCha20Poly1305 version, which requires that NONCEs are never reused with the same key. They are 64 or 96 bits long, depending on the version. Given that we have random input in the key, and we don't have too many messages in total (Most of the literature talks about encrypting individual network packages), *I personally* think we should be fine with generating it randomly, even if we're not using XChaCha20Poly1305. But I'm also aware that those are famous last words.
  * This step will yield us the NONCE, the CIPHERTEXT and an AUTHENTICATION_TAG.
4. Store a a record `(IES_DATA, NONCE, DATE, CIPHERTEXT, AUTHENTICATION_TAG)`

I will edit this post as soon as I can with a decryption concept.

**Comment by ImmoStanke** [comment]
I think it's still unclear what scenarios you want to protect against. I'll attempt to make some thoughts in that direction, maybe that helps clarify.
1. There needs to be one person or a set of people who can export the data to the authorities. This person or set of people need full read access.
2. The app needs one-time write access.
3. The app or a cron-job needs read-access on the date in order to do deletion based on time
4. We want to prevent everyone except the people authorized to export from reading anything other than date
5. We want to prevent everyone from tampering records

5. Is really hard to do unless you use blockchain technology. So you can say you'd like to restrict tampering possibility as much as reasonably possible. In that case you just prevent tampering by anyone except for people who have unrestricted access to the server.

I think for both 4 and 5, the big questions are: Who has access to the server, and thus the db? Given that there is not read functionality in the app, these are the only people who can access the data. If the set of people who have access to the server is the same as the people who are authorized to do an export, you're basically fine already.

Consider 4 (secrecy)
If you imagine an attack where someone gains access to the server who is unauthorized, you would want to have a private key that's not stored on the server, and encrypt all data with a public key that the attacker anyways gains access once they have sufficient server access to dump memory. The only time you ever use the private key to decrypt is when you actually need to do a data export to comply with a request from the authorities. You could even destroy the private key if you get a request that isn't related to corona.
This does give defense in depth in case the server is compromised.

Consider 5 (integrity): 
Public-key crypto will not help you if you want to guard against attackers that have access to the server/ram. Because you need the private key to sign. So an attacker can always sign with the latest public key. But if you only care about the date, you can rotate the private key frequently, and delete it, and only keep the public key around. Now you only need to guard against tampering of the public keys. So you shifted the problem to those, and you can export them e.g. by sending emails or writing them to an off-server log. But you might as well do the same with the database transaction log to achieve the same goal.

So I would propose to do only encryption and not care about integrity. Integrity only becomes an issue if the data is abuseable, which can be prevented by encryption as long as you trust the key holders (which you need to do in any case). If the data is used for it's intended purpose, the worst that can happen is that a) someone who is at risk of corona doesn't get notified or b) someone who's not actually at risk does get notified. Maybe the worst you can do is retroactively enter a person who has corona in order to cause unnecessary panic and potentially damage the app-runners' reputation. This requires tampering the date so it does require server access since the date is determined by the backend on write iirc. You could spot this by keeping backups of the transaction log or even just regular db backups, off server out of reach of the attacker.

So if you only want to do encryption, things become much simpler and you can for example use https://pynacl.readthedocs.io/en/stable/public/#nacl-public-sealedbox after all.

**Comment by voidus** [comment]
Yes, I agree that all of the fancy stuff I wrote is a bit overboard. Putting everything in sealedboxes should be fine for us.
